src/platform/OpenThread/GenericThreadBorderRouterDelegate.cpp

```diff
@@ -0,0 +1,213 @@
+}
+
+CHIP_ERROR GenericThreadBorderRouterDelegate::GetDataset(Thread::OperationalDataset & dataset, DatasetType type)
+{
+    otInstance * otInst = DeviceLayer::ThreadStackMgrImpl().OTInstance();
+    VerifyOrReturnError(otInst, CHIP_ERROR_INCORRECT_STATE);
+
+    ScopedThreadLock threadLock;
+    otError otErr = OT_ERROR_NONE;
+    otOperationalDatasetTlvs datasetTlvs;
+    if (type == DatasetType::kActive)
+    {
+        otErr = otDatasetGetActiveTlvs(otInst, &datasetTlvs);
+    }
+    else
+    {
+        otErr = otDatasetGetPendingTlvs(otInst, &datasetTlvs);
+    }
+    if (otErr == OT_ERROR_NONE)
+    {
+        return dataset.Init(ByteSpan(datasetTlvs.mTlvs, datasetTlvs.mLength));
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR GenericThreadBorderRouterDelegate::SetActiveDataset(const Thread::OperationalDataset & activeDataset,
+                                                               ActivateDatasetCallback * callback)
+{
+    CHIP_ERROR err = BackupActiveDataset();
+    if (err == CHIP_NO_ERROR)
+    {
+        err = DeviceLayer::ThreadStackMgrImpl().AttachToThreadNetwork(activeDataset, nullptr);
+    }
+    if (err == CHIP_NO_ERROR)
+    {
+        mCallback = callback;
+    }
+    return err;
+}
+
+void GenericThreadBorderRouterDelegate::OnPlatformEventHandler(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
+{
+    GenericThreadBorderRouterDelegate * delegate = reinterpret_cast<GenericThreadBorderRouterDelegate *>(arg);
+    if (delegate && delegate->mCallback)
+    {
+        if (event->Type == DeviceLayer::DeviceEventType::kThreadConnectivityChange &&
+            event->ThreadConnectivityChange.Result == DeviceLayer::kConnectivity_Established)
+        {
+            delegate->mCallback->OnActivateDatasetComplete(CHIP_NO_ERROR);
+            // Delete Failsafe Keys after activating dataset is completed
+            DeviceLayer::PersistedStorage::KeyValueStoreMgr().Delete(kFailsafeThreadDatasetTlvsKey);
+            delegate->mCallback = nullptr;
+        }
+    }
+}
+
+CHIP_ERROR GenericThreadBorderRouterDelegate::BackupActiveDataset()
+{
+    // If active dataset is already backed up, return with no error
+    CHIP_ERROR err = DeviceLayer::PersistedStorage::KeyValueStoreMgr().Get(kFailsafeThreadDatasetTlvsKey, nullptr, 0);
+    if (err == CHIP_NO_ERROR || err == CHIP_ERROR_BUFFER_TOO_SMALL)
+    {
+        return CHIP_NO_ERROR;
+    }
+    GetDataset(mStagingDataset, DatasetType::kActive);
+    ByteSpan dataset = mStagingDataset.AsByteSpan();
+    return DeviceLayer::PersistedStorage::KeyValueStoreMgr().Put(kFailsafeThreadDatasetTlvsKey, dataset.data(), dataset.size());
+}
+
+CHIP_ERROR GenericThreadBorderRouterDelegate::CommitActiveDataset()
+{
+    // Delete Failsafe Key when committing.
+    DeviceLayer::PersistedStorage::KeyValueStoreMgr().Delete(kFailsafeThreadDatasetTlvsKey);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR GenericThreadBorderRouterDelegate::RevertActiveDataset()
```

Comment:
So what under this function guarantees there will be no OnActivateDatasetComplete() callback?  What if the relevant `kThreadConnectivityChange` event has already been queued?

As in: this is not implementing the documented delegate API.

Reply:
Change it to, when the device receive a setactivedataset command, it will generate a random number, save it and pass it to the delegate, and the OnActivateDatasetComplete needs to check the random number. so that when the second setactivedataset command comes the saved random number will changed so that we will not call OnActivateDatasetComplete for the first command.